### PR TITLE
Fix for: CreateItem & UpdateItem Firing

### PR DESCRIPTION
**Pull request by omerson**
'Max call stack size exceede' using node-ews library.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
-  "name": "soap",
+  "name": "soap-noendlessloop",
```
Comment (herom) [inline package.json]: please revert the changes in `package.json` :smile:

Comment (omerson) [inline package.json]: I just did :)

**Comment by herom** [comment]
Please, don't just revert the changes to the package name in `package.json` but the whole changes made to the `package.json` file.

File: package.json [hunk 2]
```diff
@@ -17,21 +17,22 @@
...
   },
   "repository": {
     "type": "git",
-    "url": "https://github.com/milewise/node-soap.git"
+    "url": "git+https://github.com/omerson/node-soap"
   },
   "main": "./index.js",
   "directories": {
-    "lib": "./lib"
+    "test": "test"
```
Comment (herom) [inline package.json]: what's the purpose of this change?

File: package.json [hunk 3]
```diff
@@ -17,21 +17,22 @@
...
   },
   "scripts": {
-    "toc": "./node_modules/.bin/doctoc Readme.md --github --maxlevel 3",
+    "toc": "doctoc Readme.md --github --maxlevel 3",
```
Comment (herom) [inline package.json]: why did you change the script command? this would implicit that the user has `doctoc` installed globally - otherwise it would work with the `doctoc` lib installed in `soap`s `node_modules/` directory

File: package.json [hunk 4]
```diff
@@ -55,5 +56,9 @@
     "semver": "~5.0.3",
     "should": "~3.3.0",
     "timekeeper": "~0.0.4"
-  }
+  },
+  "bugs": {
+    "url": "https://github.com/omerson/node-soap/issues"
```
Comment (herom) [inline package.json]: please revert

File: package.json [hunk 5]
```diff
@@ -55,5 +56,9 @@
...
+  },
+  "homepage": "https://github.com/omerson/node-soap#readme"
```
Comment (herom) [inline package.json]: please revert

File: package.json [hunk 6]
```diff
@@ -17,21 +17,22 @@
     "sax": ">=0.6",
     "selectn": "^0.9.6",
     "strip-bom": "~0.3.1",
-    "xml-crypto": "~0.8.0"
+    "xml-crypto": "~0.8.0",
+    "ursa": "0.8.5 || >=0.9.4"
```
Comment (herom) [inline package.json]: `ursa` was made an `optionalDependency` as most of the users had problems with the library. please revert the change in the `dependencies` list as it is listed under `optionalDependencies` which should be sufficient.

**Comment by omerson** [comment]
package.json reverted

**Comment by herom** [comment]
There are 8 failing tests shown over at Travis CI - it would be awesome if you could fix them so that we can merge your PR :+1:

**Comment by MMRandy** [comment]
+1

**Comment by herom** [comment]
Were meant to be made on the `soap` repo @omerson ?

File: lib/wsdl.js [hunk 7]
```diff
@@ -2132,17 +2139,21 @@ function open_wsdl(uri, options, callback) {
...
-        callback(err);
-      }
-      else {
-        wsdl = new WSDL(definition, uri, options);
-        WSDL_CACHE[ uri ] = wsdl;
-        wsdl.WSDL_CACHE = WSDL_CACHE;
-        wsdl.onReady(callback);
-      }
-    });
+    if (typeof uri !== "undefined") {
```
Comment (herom) [inline lib/wsdl.js]: why just check for `=== 'undefined'`? this would fail if `uri` is `null`... wouldn't it be better to check for `typeof uri === 'string'` in order to prevent the `fs.readFile()` to fail?

File: package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "soap",
-  "version": "0.16.0",
+  "name": "soap-noendlessloop",
```
Comment (herom) [inline package.json]: please revert your changes in `package.json` altogether in order to get your PR merged :+1:

File: lib/wsdl.js [hunk 9]
```diff
@@ -2132,17 +2139,21 @@ function open_wsdl(uri, options, callback) {
...
+        }
+        else {
+          wsdl = new WSDL(definition, uri, options);
+          WSDL_CACHE[ uri ] = wsdl;
+          wsdl.WSDL_CACHE = WSDL_CACHE;
+          wsdl.onReady(callback);
+        }
+      });
+    } else {
+      console.log("URI undefiend: " + uri);
```
Comment (herom) [inline lib/wsdl.js]: please use `debug()` instead of `console.log()`

File: lib/wsdl.js [hunk 10]
```diff
@@ -1940,15 +1940,22 @@ WSDL.prototype.findChildSchemaObject = function(parameterTypeObj, childName) {
     } else {
       childNsURI = this.definitions.xmlns[typeInfo.prefix];
     }
-    var typeDef = this.findSchemaType(typeInfo.name, childNsURI);
-    if (typeDef) {
-      return this.findChildSchemaObject(typeDef, childName);
+    if(typeof childNsURI !== 'undefined' && childNsURI !== null){
+      if ((childNsURI.match("http://schemas.microsoft.com") == null) || isInsideSchemaMS) {
```
Comment (herom) [inline lib/wsdl.js]: please use always the "typesafe" equals operator `===` instead of `==`

File: lib/wsdl.js [hunk 11]
```diff
@@ -1940,15 +1940,22 @@ WSDL.prototype.findChildSchemaObject = function(parameterTypeObj, childName) {
...
+        var typeDef = this.findSchemaType(typeInfo.name, childNsURI);
+        if (typeDef) {
+          if (childNsURI.match("http://schemas.microsoft.com") != null)
```
Comment (herom) [inline lib/wsdl.js]: please use curly braces `{}` for `if` statements

File: package.json (hunk 8 above)
Comment (omerson) [inline package.json]: @herom thank you very much on suggestions I have pushed the changes

File: lib/wsdl.js [hunk 12]
```diff
@@ -2139,7 +2141,7 @@ function open_wsdl(uri, options, callback) {
   var wsdl;
   if (!/^https?:/.test(uri)) {
     debug('Reading file: %s', uri);
-    if (typeof uri !== "undefined") {
+    if (typeof uri !== "string") {
```
Comment (herom) [inline lib/wsdl.js]: I guess this should read `typeof uri === 'string'` :smiley:

Comment (omerson) [inline lib/wsdl.js]: Sorry for taking you time. Thanks again for the review.

**Comment by herom** [comment]
Thanks for your patience and frequent updates @omerson - we appreciate it a lot :+1:

LGTM - but I'd need you to provide at least 1 test for your addition so that we can assure your fix isn't going to be overwritten accidentially in future PRs.

Also, I'd love to hear @jsdevel 's opinon :smiley_cat:

**Comment by omerson** [comment]
Thanks @herom please advise on writing test, since I am new to github collaboration. Could you please provide some reading or example.

**Comment by herom** [comment]
The best test examples are found in our [request-response-samples](https://github.com/vpulim/node-soap/tree/master/test/request-response-samples) section where you can provide a small sample of your `wsdl` and expected parsed `json` results - please look through these examples as they are the best source for reproducible test results :+1:

File: package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "soap-noendlessloop",
-  "version": "0.16.3",
+  "name": "soap",
```
Comment (jsdevel) [inline package.json]: please revert these changes

Comment (omerson) [inline package.json]: yes sorry

**Comment by eino-makitalo** [comment]
Yes this fixes my problem too. Please can you accept this  sooner to production version.

**Comment by ate-f** [comment]
This happens for other WSDL files as well it seems: I'm getting it for the Salesforce Marketing Cloud wsdl: https://webservice.exacttarget.com/etframework.wsdl

**Comment by jsdevel** [comment]
> Yes this fixes my problem too. Please can you accept this sooner to production version.

As @herom mentioned earlier we need some request/response samples added to the test suite.  Merging a PR in haste to fix an issue is counterproductive if the code hasn't been tested and causes other unwanted issues, or if it breaks with future changes and we're unaware because there were no tests to verify the fix.

**Comment by herom** [comment]
Just for future-reference: we also have a [CONTRIBUTING.md](https://github.com/vpulim/node-soap/blob/master/CONTRIBUTING.md) file where we set a few "rules & standards", also for [Submitting a Pull Request](https://github.com/vpulim/node-soap/blob/master/CONTRIBUTING.md#submitting-a-pull-request). 

The reason why we ask for this is best summarized in @jsdevel's post above

>Merging a PR in haste to fix an issue is counterproductive if the code hasn't been tested and causes other unwanted issues, or if it breaks with future changes and we're unaware because there were no tests to verify the fix.

Perhaps it's a good idea to bring this 1 sentence into the first chapter of our `README.md` so that more people are aware of our "PR policy"?

**Comment by jsdevel** [comment]
closing due to inactivity
